File: C11/백준/Gold/1005. ACM Craft/ACM Craft.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 1001

int D[MAX], dp[MAX], indegree[MAX];
int adj[MAX][MAX];

int main() {
    int T;
    scanf("%d", &T);

    while (T--) {
        int N, K;
        scanf("%d %d", &N, &K);

        for (int i = 1; i <= N; i++) {
            scanf("%d", &D[i]);
            dp[i] = 0;
            indegree[i] = 0;
            for (int j = 1; j <= N; j++) adj[i][j] = 0;
        }

        for (int i = 0; i < K; i++) {
            int X, Y;
            scanf("%d %d", &X, &Y);
            adj[X][Y] = 1;
            indegree[Y]++;
        }

        int W;
        scanf("%d", &W);

        int queue[MAX], front = 0, rear = 0;

        for (int i = 1; i <= N; i++) {
            dp[i] = D[i];
            if (indegree[i] == 0) queue[rear++] = i;
        }

        while (front < rear) {
            int now = queue[front++];

            for (int next = 1; next <= N; next++) {
                if (adj[now][next]) {
                    if (dp[next] < dp[now] + D[next]) {
                        dp[next] = dp[now] + D[next];
                    }
                    indegree[next]--;
                    if (indegree[next] == 0) queue[rear++] = next;
                }
            }
        }

        printf("%d\n", dp[W]);
    }

    return 0;
}

File: Java/백준/Gold/1005. ACM Craft/ACM Craft.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());

            int[] D = new int[N + 1];
            int[] indegree = new int[N + 1];
            List<Integer>[] graph = new ArrayList[N + 1];
            for (int i = 1; i <= N; i++) graph[i] = new ArrayList<>();

            st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= N; i++) D[i] = Integer.parseInt(st.nextToken());

            for (int i = 0; i < K; i++) {
                st = new StringTokenizer(br.readLine());
                int X = Integer.parseInt(st.nextToken());
                int Y = Integer.parseInt(st.nextToken());
                graph[X].add(Y);
                indegree[Y]++;
            }

            int W = Integer.parseInt(br.readLine());

            int[] dp = new int[N + 1];
            Queue<Integer> q = new LinkedList<>();

            for (int i = 1; i <= N; i++) {
                dp[i] = D[i];
                if (indegree[i] == 0) q.add(i);
            }

            while (!q.isEmpty()) {
                int now = q.poll();
                for (int next : graph[now]) {
                    dp[next] = Math.max(dp[next], dp[now] + D[next]);
                    indegree[next]--;
                    if (indegree[next] == 0) q.add(next);
                }
            }

            System.out.println(dp[W]);
        }
    }
}